pkg/tlsx/openssl/openssl.go

```diff
@@ -51,94 +41,61 @@ func (c *Client) ConnectWithOptions(hostname, ip, port string, options clients.C
 		address = net.JoinHostPort(hostname, port)
 	}
 
-	opensslCtx, err := openssl.NewCtxWithVersion(openssl.AnyVersion)
-	if err != nil {
-		return nil, err
+	// Note: CLI options are omitted if given value is empty
+	opensslOptions := Options{
+		Address:    address,
+		ServerName: options.SNI,
+		Protocol:   getProtocol(options.VersionTLS),
+		CAFile:     c.options.CACertificate,
+		Cipher:     validateCiphers(options.Ciphers...),
 	}
-	opensslCtx.SetVerifyMode(openssl.VerifyNone)
 
-	if c.options.Timeout > 0 {
-		opensslCtx.SetTimeout(time.Duration(c.options.Timeout) * time.Second)
+	if opensslOptions.ServerName == "" {
+		// If there are multiple VHOST openssl returns errors unless hostname is specified (ex: projectdiscovery.io)
+		opensslOptions.ServerName = hostname
 	}
 
-	if len(c.options.Ciphers) > 0 {
-		if err := opensslCtx.SetCipherList(strings.Join(c.options.Ciphers, ",")); err != nil {
-			return nil, errors.Wrap(err, "could not set ciphers")
-		}
+	// timeout cannot be zero(If GOOS= windows it should be on average 3)
+	// this timeout will be used by os.exec context
+	if c.options.Timeout < 3 {
+		c.options.Timeout = 3
 	}
 
-	if c.options.CACertificate != "" {
-		caCert, err := ioutil.ReadFile(c.options.CACertificate)
-		if err != nil {
-			return nil, errors.Wrap(err, "could not read ca certificate")
-		}
-		caStore := opensslCtx.GetCertificateStore()
-		err = caStore.LoadCertificatesFromPEM(caCert)
+	// validate dialer before using
+	if c.dialer == nil {
+		var err error
+		c.dialer, err = fastdialer.NewDialer(fastdialer.DefaultOptions)
 		if err != nil {
-			return nil, errors.Wrap(err, "could not add certificate to store")
+			return nil, fmt.Errorf("openssl: failed to create new dialer %v", c.dialer)
```

Comment:
shouldn't be `err` instead of `c.dialer`? I also think we should use `errors.Wrap` or `errors.Wrapf`